libs-scala/ports/src/main/scala/com/digitalasset/ports/PortFiles.scala

```diff
@@ -36,9 +36,10 @@ object PortFiles {
     }
 
   private def writeUnsafe(path: Path, port: Port): Unit = {
-    import java.nio.file.StandardOpenOption.CREATE_NEW
     val lines: java.lang.Iterable[String] = List(port.value.toString).asJava
-    val created = Files.write(path, lines, CREATE_NEW)
+    val tmpFile = Files.createTempFile("portfile", "")
+    Files.write(tmpFile, lines)
+    val created = Files.move(tmpFile, path)
```

Comment:
There is an `ATOMIC_MOVE` flag, maybe we should use that and fall back to a non-atomic move if we have to? I’m not quite sure if not using that flag is equivalent to that or if it will always be non-atomic.

Reply:
I was thinking the same thing when I read the docs. Not sure if ATOMIC_MOVE is the default or if we need to set it explicitly and deal with the possibility that the atomic move is not possible on a system.

Reply:
I guess if the docs are unclear, trying the atomic move first is the safe option? :shrug:

Reply:
From the [docs](https://docs.oracle.com/javase/8/docs/api/java/nio/file/Files.html#move-java.nio.file.Path-java.nio.file.Path-java.nio.file.CopyOption...-):

> If the target file exists then it is implementation specific if the existing file is replaced or this method fails by throwing an IOException.

I know this works as you'd expect on Linux but I have no idea about other systems.

Reply:
Based on digging into this a little when writing _ledger-on-filesystem_, I came to the conclusion that on Linux and macOS, it probably does nothing _except_ when you try and move across file systems, in which case it errors because you can't do an atomic move across file systems.

On Windows, I _think_ it does nothing, but I'm not sure.

Reply:
That seems like a very questionable API. Why does requesting an atomic move change whether it overwrites a file or not :scream_cat:

Reply:
Let's set the flag to be on the safe side then. I wouldn't bother with the additional error case for now though.

Reply:
So I added ATOMIC_MOVE, which seems to work just fine on my local system. But a lot of tests are failing now on CI. Does anybody have an idea what's going on?

Reply:
> java.nio.file.AtomicMoveNotSupportedException: /home/vsts/.cache/bazel/_bazel_vsts/d5c8b3f1185416e5207d53317d3a7876/sandbox/linux-sandbox/667/execroot/com_github_digital_asset_daml/_tmp/e05013f2f0106511efb2381068335f20/portfile3984526891245692342 -> PortFilesSpec-8504083e-142d-42f8-8bee-d6495e77540d.dummy: Invalid cross-device link

Looks like the Bazel-allocated temporary directory isn't on the same volume.

I suggest just turning it off again and calling it best-effort.

Reply:
Looks like it's ont that easy. Without `ATOMIC_MOVE` tests fail for me locally, while they pass with the flag set, but don't pass CI...

Reply:
@robin-da how does it fail locally? Presumably not with the same error?

Reply:
tests fail with the `portfile not written in time` error. I try now to add a fallback for when `ATOMIC_MOVE` is not supported.

Reply:
huh that seems very odd. Which test is that? I can try to reproduce.

Reply:
you can try assistant integration tests

Reply:
This is expected. daml-helper will use an already existing file as the port file. Here’s the fix:

```
diff --git a/daml-assistant/daml-helper/src/DA/Daml/Helper/Start.hs b/daml-assistant/daml-helper/src/DA/Daml/Helper/Start.hs
index 330954788..7386faa10 100644
--- a/daml-assistant/daml-helper/src/DA/Daml/Helper/Start.hs
+++ b/daml-assistant/daml-helper/src/DA/Daml/Helper/Start.hs
@@ -100,7 +100,8 @@ navigatorURL :: NavigatorPort -> String
 navigatorURL (NavigatorPort p) = "http://localhost:" <> show p
 
 withSandbox :: SandboxClassic -> SandboxPortSpec -> [String] -> (Process () () () -> SandboxPort -> IO a) -> IO a
-withSandbox (SandboxClassic classic) portSpec extraArgs a = withTempFile $ \portFile -> do
+withSandbox (SandboxClassic classic) portSpec extraArgs a = withTempDir $ \dir -> do
+    let portFile = dir </> "port-file"
     let sandbox = if classic then "sandbox-classic" else "sandbox"
     let args = [ sandbox
                , "--port", show (fromSandboxPortSpec portSpec)
```

However this gets us back to `ATOMIC_MOVE` overwriting existing files. I think we just need to not use `ATOMIC_MOVE` and hope that at least on Linux it’s still atomic. I don’t see a way to force an atomic move while not also overwriting the file (any kind of exists check is racy).

Reply:
ok that fixes it. No more `ATOMIC_MOVE` and :crossed_fingers: .